Objective-C/CBLDictionary.mm

```diff
@@ -321,27 +390,45 @@ - (NSArray*) allKeys {
 
 
 - (void) setChanged {
+    // Has it own _changedLock instead of using the _lock to avoid deadlock when encoding
+    // the object into fleece and at the same time there are some changes in a child
+    // dictionary or array by another thread. There is one side effect of doing this is that
+    // the changed status could be NO when one thread is trying to get the changed status
+    // but the other thread is modifying the object at the same time.
+    [_changedLock lock];
     if (!_changed) {
         _changed = YES;
         [self notifyChangeListeners];
```

Comment:
It's generally dangerous to call into code you don't control, while holding a lock.
In this case I think deadlocks are possible:
* From -setChanged, a change listener calls back into this object, which will then try to acquire `_lock`.
* Meanwhile, another thread is in a mutating method of this object that already has `_lock` and then calls into -setChanged to acquire `_changedLock`.